### ISSUE TITLE
Save memory checkpoints

### ISSUE THREAD

**Issue by andreasceid** [Curt-Park/rainbow-is-all-you-need#49]
Hello,

I read the tutorial on rainbow `08.rainbow.ipynb` and I really liked it. I need some help coding a method for efficiently saving the memory of the DQN (the `ReplayBuffer` object and the `PrioritizedReplayBuffer` object). I used numpy's `savez` method to save the `ReplayBuffer` like this:

```py
class ReplayBuffer:
    """A simple numpy replay buffer."""

    def __init__(
        self, 
        obs_dim: tuple, 
        size: int, 
        save_dir: str,
        batch_size: int = 32, 
        n_step: int = 1, 
        gamma: float = 0.99
    ):
        self.obs_buf = numpy.zeros((size, *obs_dim), dtype=numpy.float32)
        self.next_obs_buf = numpy.zeros((size, *obs_dim), dtype=numpy.float32)
        self.acts_buf = numpy.zeros(size, dtype=numpy.int64)
        self.rews_buf = numpy.zeros(size, dtype=numpy.float32)
        self.done_buf = numpy.zeros(size, dtype=numpy.uint8)
        self.max_size, self.batch_size = size, batch_size
        self.ptr, self.size, = 0, 0

        self.save_dir = save_dir

        # for N-step Learning
        self.n_step_buffer = deque(maxlen=n_step)
        self.n_step = n_step
        self.gamma = gamma

    def store(
        self, 
        obs: numpy.ndarray, 
        act: numpy.ndarray, 
        rew: float, 
        next_obs: numpy.ndarray, 
        done: bool,
    ) -> Tuple[numpy.ndarray, numpy.ndarray, float, numpy.ndarray, bool]:
        transition = (obs, act, rew, next_obs, done)
        self.n_step_buffer.append(transition)

        # single step transition is not ready
        if len(self.n_step_buffer) < self.n_step:
            return ()
        
        # make a n-step transition
        rew, next_obs, done = self._get_n_step_info(self.n_step_buffer, self.gamma)
        obs, act = self.n_step_buffer[0][:2]
        
        self.obs_buf[self.ptr] = obs
        self.next_obs_buf[self.ptr] = next_obs
        self.acts_buf[self.ptr] = act
        self.rews_buf[self.ptr] = rew
        self.done_buf[self.ptr] = done
        self.ptr = (self.ptr + 1) % self.max_size
        self.size = min(self.size + 1, self.max_size)
        
        return self.n_step_buffer[0]

    def sample_batch(self) -> Dict[str, numpy.ndarray]:
        idxs = numpy.random.choice(self.size, size=self.batch_size, replace=False)

        return dict(
            obs=self.obs_buf[idxs],
            next_obs=self.next_obs_buf[idxs],
            acts=self.acts_buf[idxs],
            rews=self.rews_buf[idxs],
            done=self.done_buf[idxs],
            # for N-step Learning
            indices=idxs,
        )
    
    def sample_batch_from_idxs(
        self, idxs: numpy.ndarray
    ) -> Dict[str, numpy.ndarray]:
        # for N-step Learning
        return dict(
            obs=self.obs_buf[idxs],
            next_obs=self.next_obs_buf[idxs],
            acts=self.acts_buf[idxs],
            rews=self.rews_buf[idxs],
            done=self.done_buf[idxs],
        )
    
    def _get_n_step_info(
        self, n_step_buffer: Deque, gamma: float
    ) -> Tuple[numpy.int64, numpy.ndarray, bool]:
        """Return n step rew, next_obs, and done."""
        # info of the last transition
        rew, next_obs, done = n_step_buffer[-1][-3:]

        for transition in reversed(list(n_step_buffer)[:-1]):
            r, n_o, d = transition[-3:]

            rew = r + gamma * rew * (1 - d)
            next_obs, done = (n_o, d) if d else (next_obs, done)

        return rew, next_obs, done
    
    def save_buffer(self):
        save_path = self.save_dir / f"xp_buffer-{self.ptr}-{self.max_size}.npz"
        numpy.savez_compressed(save_path, state_memory=self.obs_buf, next_state_memory=self.next_obs_buf,
                               action_memory=self.acts_buf, reward_memory=self.rews_buf, terminal_memory=self.done_buf)
        print(f"Memory Buffer saved to {save_path} of size {self.ptr} and total capacity {self.max_size}")

    def load_buffer(self, chkpt_path: str):
        if not chkpt_path.exists():
            raise ValueError(f"{chkpt_path} does not exist")

        path = os.path.normpath(chkpt_path)
        chkpt = numpy.load(chkpt_path)
        tokens = path.split('-')

        self.obs_buf = chkpt['state_memory']
        self.acts_buf = chkpt['action_memory']
        self.rews_buf = chkpt['reward_memory']
        self.done_buf = chkpt['terminal_memory']
        self.next_obs_buf = chkpt['next_state_memory']
        self.ptr = int(tokens[1])
        self.max_size = int(os.path.splitext(tokens[2])[0])
        
        print(f"Loading buffer at {chkpt_path} of size {self.ptr} and total capacity {self.max_size}")

    def __len__(self) -> int:
        return self.size
```

I don't know how to save the object created from `PrioritizedReplayBuffer` class. Any help is appreciated :)

**Comment by Curt-Park** [comment]
You should save the additional properties for `PrioritizedReplayBuffer`.

```
self.max_priority
self.tree_ptr
self.alpha
self.sum_tree
self.min_tree
```

**Comment by Curt-Park** [comment]
or simply you can just dump the memory itself.

```python
import pickle 

# file dump
with open("memory.pkl", "wb") as fh:
    pickle.dump(self.memory, fh)
    
# file load
with open("memory.pkl", "rb") as fh:
    memory = pickle.load(fh)
```

**Comment by andreasceid** [comment]
This creates a pickle of size 21 GB given a total tree capacity of 1e5 elements. The `numpy` implementation that I attached above creates a compressed file of size 185 MB, which is way more efficient. I think that I should be able to save the `Priority Experience Replay Buffer` using approximately the same amount of memory.

**Comment by Curt-Park** [comment]
For more efficient compression, one possible way would be to change the type of `self.tree` in `SegmentTree` from `list` to `np.ndarray` and save it using `numpy.savez_compressed`.

https://github.com/Curt-Park/rainbow-is-all-you-need/blob/master/segment_tree.py#L34

**Comment by andreasceid** [comment]
I've got news. Unfortunately, I performed the comparison incorrectly. It seems that numpy saves only the elements changed in memory. So for example, if I have inserted 20k elements into the memory, it will only save those 20k elements, hence 350 MB. However, if I fill the whole memory up with elements, then the file covers 18 GB of memory capacity. Pickle doesn't use more memory. It directly saves all elements of an object in the memory, hence the file exported will always be around the same size (21 GB). So I'm sorry for the (huge) inaccuracy before.
